### PR TITLE
Updated install script to work with openDsh

### PR DESCRIPTION
**Pull request by rhysmorgan134**
Script now beings in dash, then clones the 3 required directories as peers of dash. Then runs through the process to install them all. Added arguments to choose what part of the script will run

* no options will run a full install of all of the below
* --deps - will install dependencies
* --aasdk - will install aasdk
* --gstreamer - will install gstreamer
* --openauto - will install open auto
* --dash - will install dash

example will only install dependencies, openauto and dash.
```./install.sh --deps --openauto --dash```

### CODE REVIEW

**Comment by egisz** [comment]
HI, 
Installing on clean / updated Raspbian Buster, I get error:
````
[100%] Linking CXX executable ../bin/autoapp
/usr/bin/ld: cannot find -lomx
collect2: error: ld returned 1 exit status
make[2]: *** [autoapp/CMakeFiles/autoapp.dir/build.make:188: bin/autoapp] Error 1
make[1]: *** [CMakeFiles/Makefile2:270: autoapp/CMakeFiles/autoapp.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
make err`or check logs
````

**Comment by rhysmorgan134** [comment]
@egisz apologies, it was still looking at the original fork, updated now to look at one with the fix, should work now

**Comment by egisz** [comment]
Thanks @rhysmorgan134, now it works. 
Probably it would be best to to create another PR for openauto Repo with your fixes. 
And then fix this script to point to original Repo. What do you think?

**Comment by rhysmorgan134** [comment]
> Thanks @rhysmorgan134, now it works.
> Probably it would be best to to create another PR for openauto Repo with your fixes.
> And then fix this script to point to original Repo. What do you think?

Yeah, the issue was brought up in slack, @icecube45 mentioned that its most likely something else going on. I'm not great with C++, so hopefully one of the guys can look into it, (I still need to triple check that all dependencies are correct and its not an issue on my side thats not showing as its not doing an OMX build)

This PR is in draft, I won't submit it until this is resolved in which way it requires.

File: install.sh [hunk 1]
```diff
@@ -1,7 +1,69 @@
 #!/bin/bash
 
+#repo addresses
+aasdkRepo="https://github.com/OpenDsh/aasdk"
+gstreamerRepo="git://anongit.freedesktop.org/gstreamer/qt-gstreamer"
+openautoRepo="https://github.com/rhysmorgan134/openauto.git"
```
Comment (stefan-sherwood) [inline install.sh]: This will change at some point, I trust.

Comment (rhysmorgan134) [reply install.sh]: For sure, I won't be submitting it until it's pointing back at the main repo, got some things to test from slack which hopefully will highlight the cause of the issue.

File: install.sh [hunk 2]
```diff
@@ -1,7 +1,69 @@
...
+#set default arguments
+deps=false
+aasdk=false
+gstreamer=false
+openauto=false
+dash=false
+perms=false
+
+if [ $# -gt 0 ]; then
+    while [ "$1" != "" ]; do
```
Comment (stefan-sherwood) [inline install.sh]: These two lines can be condensed to

`while [ $# -gt 0 ]; do`

File: install.sh [hunk 3]
```diff
@@ -22,86 +84,342 @@ deps=(
 "librtaudio6"
 "libkf5bluezqt-dev"
 "libtag1-dev"
+"qml-module-qtquick2"
+"doxygen"
```
Comment (stefan-sherwood) [inline install.sh]: I don't remember seeing this as a needed library and I don't see any reference to it in the code.

File: install.sh [hunk 4]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+"qml-module-qtquick*"
```
Comment (stefan-sherwood) [inline install.sh]: Seems a bit strange that we would need both `qml-module-qtquick` and `qml-module-qtquick2`

File: install.sh [hunk 5]
```diff
@@ -22,86 +84,342 @@ deps=(
...
-  exit
-else
-  echo "made ok"
+###############################  dependencies  #########################
+if [ "$deps" = true ]; then
+  #loop through dependencies and install
+	echo "installing dependencies"
+	echo "Running apt update"
+	sudo apt update
+  for app in ${dependencies[@]}; do
```
Comment (stefan-sherwood) [inline install.sh]: Is there a reason to install each of these independently? It is much slower than doing it all at once.

File: install.sh [hunk 6]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+    echo "installing: " $app
+    sudo apt -qq -o=Dpkg::Use-Pty=0 install $app -y > /dev/null 2> /dev/null
+
+    if [[ $? > 0 ]]
+      then
+          echo $app " Failed to install, quitting"
```
Comment (stefan-sherwood) [inline install.sh]: FWIW, the quotation marks are optional. It works fine as

`echo $app Failed to install, quitting`

File: install.sh [hunk 7]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+          exit
+      else
+          echo $app " Installed ok"
+          echo
```
Comment (stefan-sherwood) [inline install.sh]: FWIW, this also works

`echo -e echo $app Installed ok\n`

Just a matter of taste

File: install.sh [hunk 8]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+    fi
+  done
+  echo "All dependencies installed"
   echo
+else
```
Comment (stefan-sherwood) [inline install.sh]: Also a personal preference thing but because the "if" is super long and the "else" is super short you can reverse the if statement and put this part first and then it's easier to see what's what.

File: install.sh [hunk 9]
```diff
@@ -22,86 +84,342 @@ deps=(
...
-#begin cmake
-echo "beginning cmake for raspberry pi"
-cmake -DRPI_BUILD=TRUE -DGST_BUILD=TRUE -DCMAKE_BUILD_TYPE=Release . 
-if [[ $? > 0 ]]
-  then
-    echo "Cmake error"
-  exit
+
+###############################  AASDK #########################
+if [ "$aasdk" = true ]; then
```
Comment (stefan-sherwood) [inline install.sh]: quotation marks are superfluous

File: install.sh [hunk 10]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+  cd aasdk
+
+  #beginning cmake
+  cmake -DCMAKE_BUILD_TYPE=Release .
+  if [[ $? > 0 ]]
+    then
+      echo "CMake error check logs"
+    exit
+  else
+    echo "Cmake ok"
```
Comment (stefan-sherwood) [inline install.sh]: Nitpick: capitalization

File: install.sh [hunk 11]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+    echo
+  fi
+
+  #beginning make
+  make -j2
+
+  if [[ $? > 0 ]]
```
Comment (stefan-sherwood) [inline install.sh]: Another personal preference: putting success first, condensing, e.g.

```
if [[ $? -eq 0 ]]; then
    echo -Make completed successfully
else
    echo -e Make failed with code $?\n
fi
```

File: install.sh [hunk 12]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+  echo "installing gstreamer"
+  #change to parent directory
+  cd ..
+  #clone gstreamer
+  echo "Cloning Gstreamer"
+  git clone $gstreamerRepo
+  if [[ $? > 0 ]]
+    then
+      cd qt-gstreamer
+      if [[ $? > 0 ]]
```
Comment (stefan-sherwood) [inline install.sh]: You're checking that the `cd` worked properly?

File: install.sh [hunk 13]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+    echo "cloned OK"
+    echo
+  fi
+
+  #change into newly cloned directory
+  cd qt-gstreamer
+
+  #create build directory
+  echo "creating Gstreamer build directory"
+
```
Comment (stefan-sherwood) [inline install.sh]: Remove blank line

File: install.sh [hunk 14]
```diff
@@ -22,86 +84,342 @@ deps=(
...
   else
-    echo "make ok, executable can be found ../bin/ia"
+    echo "make ok"
     echo
+  fi
 
-    #check if usb rules exist
-    echo "checking if permissions exist"
+  #make j4
+  echo "making J4"
```
Comment (stefan-sherwood) [inline install.sh]: This is confusing. This is building `qt-gstreamer`, not `J4` and the message should reflect that.

File: install.sh [hunk 15]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+  if [[ $? > 0 ]]
+    then
+      echo "install error check logs"
+    exit
+  else
+    echo "installed ok"
+    echo
+  fi
+
+  #run ldconfig
```
Comment (stefan-sherwood) [inline install.sh]: Comments that say exactly what the code says are superfluous

File: install.sh [hunk 16]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+
+
+###############################  openauto  #########################
+if [ "$openauto" = true ]; then
+  echo "installing openauto"
+  cd ..
 
-        # OPEN USB RULE, CREATE MORE SECURE RULE IF REQUIRED
-        echo "SUBSYSTEM==\"usb\", ATTR{idVendor}==\"*\", ATTR{idProduct}==\"*\", MODE=\"0660\", GROUP=\"plugdev\"" | sudo tee $FILE
+  clone openauto
```
Comment (stefan-sherwood) [inline install.sh]: Maybe it's just the diff but I think you left out an `echo` here.

File: install.sh [hunk 17]
```diff
@@ -22,86 +84,342 @@ deps=(
...
 fi
 
-#Start app
-echo "starting app"
-cd ../bin
-./ia
+
+###############################  dash  #########################
+if [ "$dash" = true ]; then
+  #loop through dependencies and install
```
Comment (stefan-sherwood) [inline install.sh]: This comment seems irrelevant to the code

File: install.sh [hunk 18]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+  fi
+
+  echo "running make"
+  make
+  if [[ $? > 0 ]]
+    then
+      echo "make error check output above"
+      exit
+    else
+      echo "make ok, executable can be found ../bin/ia"
```
Comment (stefan-sherwood) [inline install.sh]: When these messages appear the context will be lost. I would include the name of the module being built in every message.

File: install.sh [hunk 19]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+      echo
+
+      #check if usb rules exist
+      echo "checking if permissions exist"
```
Comment (stefan-sherwood) [inline install.sh]: Message is too generic

File: install.sh [hunk 20]
```diff
@@ -22,86 +84,342 @@ deps=(
...
+
+      #udev rule to be created below, change as needed
+      FILE=/etc/udev/rules.d/51-iadash.rules
+      if [[ -f "$FILE" ]]
+        then
+          echo "rules exists"
+          echo
+        else
+          sudo touch $FILE
```
Comment (stefan-sherwood) [inline install.sh]: This is unneeded. `tee` will create the file and if you do it this way and the `echo` or `tee` below fails, a subsequent run will errantly detect that the rules exist and produce a broken product.

**Review by stefan-sherwood — COMMENTED** [review]
Some of my comments can be applied throughout.

File: install.sh (hunk 2 above)
Comment (rhysmorgan134) [reply install.sh]: The reason for the outer IF is to catch whether there are no arguments supplied, which if there isn't it runs the full install

File: install.sh (hunk 3 above)
Comment (rhysmorgan134) [reply install.sh]: doxygen was to remove a warning whilst compiling, will remove

File: install.sh (hunk 2 above)
Comment (stefan-sherwood) [reply install.sh]: Now I see what you're doing but I find it a bit confusing. Before this section you set what you call "defaults" but if one executes the script without arguments it doesn't use those defaults. At the least I would put the setting of the defaults inside the `if` clause to which it solely applies.

Comment (rhysmorgan134) [reply install.sh]: renamed to initialising arguments and moved into the if statement

https://github.com/openDsh/dash/blob/a70d382506d021b17a2d9e41465acccda00c192d/install.sh#L29-L38

File: install.sh (hunk 4 above)
Comment (rhysmorgan134) [reply install.sh]: ```qml-module-qtquick*``` has been commented out, will remove once verified via fresh install

File: install.sh (hunk 5 above)
Comment (rhysmorgan134) [reply install.sh]: Now concatenate a string with the full install (is indeed much faster :) ) echo line will be removed before submission

https://github.com/openDsh/dash/blob/a70d382506d021b17a2d9e41465acccda00c192d/install.sh#L120-L129

File: install.sh (hunk 6 above)
Comment (rhysmorgan134) [reply install.sh]: All echo quotations removed

File: install.sh (hunk 7 above)
Comment (rhysmorgan134) [reply install.sh]: Updated to work like this, it does look much cleaner, had to use it like this though

```echo -e $app Installed ok '\n'```

File: install.sh (hunk 8 above)
Comment (rhysmorgan134) [reply install.sh]: Switched these around in all cases of really long statements

File: install.sh (hunk 9 above)
Comment (rhysmorgan134) [reply install.sh]: removed all cases

File: install.sh (hunk 10 above)
Comment (rhysmorgan134) [reply install.sh]: updated to CMake

File: install.sh (hunk 11 above)
Comment (rhysmorgan134) [reply install.sh]: Agreed it does read better like this, all "if" statements now updated to reflect

File: install.sh (hunk 12 above)
Comment (rhysmorgan134) [reply install.sh]: removed...

File: install.sh (hunk 13 above)
Comment (rhysmorgan134) [reply install.sh]: removed

File: install.sh (hunk 14 above)
Comment (rhysmorgan134) [reply install.sh]: Updated

File: install.sh (hunk 15 above)
Comment (rhysmorgan134) [reply install.sh]: removed

File: install.sh (hunk 16 above)
Comment (rhysmorgan134) [reply install.sh]: removed

File: install.sh (hunk 17 above)
Comment (rhysmorgan134) [reply install.sh]: removed

File: install.sh (hunk 18 above)
Comment (rhysmorgan134) [reply install.sh]: Updated all result "if..else" to include section name

File: install.sh (hunk 19 above)
Comment (rhysmorgan134) [reply install.sh]: updated

File: install.sh (hunk 20 above)
Comment (rhysmorgan134) [reply install.sh]: updated

https://github.com/openDsh/dash/blob/a70d382506d021b17a2d9e41465acccda00c192d/install.sh#L367-L377

File: install.sh [hunk 21]
```diff
@@ -22,86 +92,304 @@ deps=(
...
-  then
-    echo "make error check output above"
+  #clone aasdk
+  git clone $aasdkRepo
+  if [[ $? -eq 0 ]]; then
+    echo -e Aasdk Cloned ok '\n'
+  else
+    cd aasdk
+    if [[ $? -eq 0 ]]; then
+      git pull $aasdkRepo
```
Comment (egisz) [inline install.sh:149]: Please add check if command was successful, as script continues in case of error: 
```
pi@raspberrypi:~/dash $ ./install.sh --aasdk
skipping dependencies
fatal: destination path 'aasdk' already exists and is not an empty directory.
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Total 4 (delta 3), reused 3 (delta 3), pack-reused 0
Unpacking objects: 100% (4/4), done.
From https://github.com/OpenDsh/aasdk
 * branch            HEAD       -> FETCH_HEAD
Updating 9170cef..7cb2b41
error: Your local changes to the following files would be overwritten by merge:
        aasdk_proto/ButtonCodeEnum.proto
Please commit your changes or stash them before you merge.
Aborting
cloned OK

```

File: install.sh [hunk 22]
```diff
@@ -22,86 +92,304 @@ deps=(
...
+
+  #clone gstreamer
+  echo Cloning Gstreamer
+  git clone $gstreamerRepo
+  if [[ $? -eq 0 ]]; then
+    echo -e Gstreamer cloned OK
+  else
+    cd qt-gstreamer
+      if [[ $? -eq 0 ]]; then
+        git pull gstreamerRepo
```
Comment (egisz) [inline install.sh]: same here:  Please add check if command was successful, as script continues in case of error:

Comment (rhysmorgan134) [reply install.sh]: I'd guess it's debatable whether you would want it to stop, if you are working and have changes, you would want it to build with the changes rather then exit on commits not added?

File: install.sh [hunk 23]
```diff
@@ -1,7 +1,78 @@
...
+if grep -q "Raspbian" ${OS_RELEASE_FILE}; then
+  installArgs="-DRPI_BUILD=true"
+  isRpi=true
+else
+  installArgs=""
+  isRpi=false
+fi
+
+#check to see if there are any arguments supplied, if none are supplied run full install
+if [ $# -gt 0 ]; then
```
Comment (stefan-sherwood) [inline install.sh:33]: Definitely better. As a fun exercise, note this alternate way of parsing the command line parameters:

```
args=( "$@" )
valid_args=( deps aasdk gstreamer openauto dash perms )

for option in "${valid_args[@]}"; do
    if [[ "${args[@]}" =~ ${option} ]]; then
       declare $option=true
    fi
done

# blah blah blah

if [ ! -z "$aasdk" ]; then
    # do aasdk stuff
fi
```

This allows you to set the variables externally as environment variables instead of passing them in on the script. If you don't want that you can add `else declare $option=false`.

Comment (stefan-sherwood) [reply install.sh:33]: It also opens the door to validating that the given arguments are valid.

File: install.sh [hunk 24]
```diff
@@ -1,7 +1,78 @@
...
+    echo "   --dash           install and build dash "
+    echo
+    exit 1
+}
+
+#location of OS details for linux
+OS_RELEASE_FILE="/etc/os-release"
+
+#check if Raspian is in the file, if not set the install Args to be false
+if grep -q "Raspbian" ${OS_RELEASE_FILE}; then
```
Comment (rsjudka) [inline install.sh:24]: @robert5974 might be able to weigh in here, is the new Raspberry OS still internally named "Raspbian"?

Comment (robert5974) [reply install.sh:24]: When I check for an update it still says Raspbian Buster.  That is with a full dist-upgrade from Buster and also a clean flash using the raspberry pi flasher utility which will auto pull the latest is image.

**Comment by egisz** [comment]
You may already know, but I'd like to point out that Crankshaft contains [prebuilds scripts](https://github.com/presslab-us/prebuilts/tree/master/buildsystem) for same things. Maybe worth to reuse?
There also is qt build script which may also be useful, what do you think?

File: install.sh [hunk 25]
```diff
@@ -1,7 +1,78 @@
...
+
+#check to see if there are any arguments supplied, if none are supplied run full install
+if [ $# -gt 0 ]; then
+  #initialize all arguments to false before looping to find which are available
+  deps=false
+  aasdk=false
+  gstreamer=false
+  openauto=false
+  dash=false
+  perms=false
```
Comment (rsjudka) [inline install.sh]: i dont see `perms` being used anywhere... also dont see it in the help print

File: install.sh [hunk 26]
```diff
@@ -1,7 +1,78 @@
 #!/bin/bash
 
+#repo addresses
+aasdkRepo="https://github.com/OpenDsh/aasdk"
```
Comment (rsjudka) [inline install.sh:4]: does this (and the repo below it) not need the `.git` at the end?

File: install.sh [hunk 27]
```diff
@@ -1,7 +1,78 @@
...
+  #initialize all arguments to false before looping to find which are available
+  deps=false
+  aasdk=false
+  gstreamer=false
+  openauto=false
+  dash=false
+  perms=false
+    while [ "$1" != "" ]; do
+        case $1 in
+            --deps )           shift
```
Comment (rsjudka) [inline install.sh:42]: spacing seems kinda wonky here? tbh not too familiar with switch cases in bash tho so excuse my ignorance if this is how it should be :p

File: install.sh [hunk 28]
```diff
@@ -22,86 +93,332 @@ deps=(
...
+
+
+###############################  openauto  #########################
+if [ $openauto = false ]; then
+  echo -e skipping openauto'\n'
+else
+  echo Installing openauto
+  cd ..
+
+  #Make ilclient
```
Comment (rsjudka) [inline install.sh]: this isnt really needed anymore since we're forcing everyone to use gstreamer?

File: install.sh [hunk 29]
```diff
@@ -22,86 +93,332 @@ deps=(
...
+      exit
+    fi
+  fi
+
+  #change into aasdk folder
+  echo -e moving to aasdk '\n'
+  cd aasdk
+
+  #beginning cmake
+  cmake -DCMAKE_BUILD_TYPE=Release .
```
Comment (rsjudka) [inline install.sh]: would we also want to create a `build` directory and do the build there? Just in case something fails the entire build can just be wiped and retired (same for openauto)
